### PR TITLE
enhancement/legistar-sponsors-bodies

### PR DESCRIPTION
**Pull request by dphoria**
### Link to Relevant Issue

https://github.com/CouncilDataProject/cdp-scrapers/issues/3

#### Enhancement ####
Additional Legistar API queries in `legistar_utils.get_legistar_events_for_timespan()`
- GET Bodies -> `EventIngestionModel.body`, `Role.body`
- GET OfficeRecords -> `Person.roles`
- GET Sponsors -> `Matter.sponsors`

The probability of change to a given person/body in Legistar during the lifetime of a single call to `get_legistar_events_for_timespan()` is _very_ low. `legistar_utils.known_persons` and `legistar_utils.known_bodies` are used to query a given body/person just once, during a single to call to `get_legistar_events_for_timespan()`.

#### Bugfix ####
I noticed that inclusion of time zone information in the `datetime` fields in `IngestionModel`s got broken, such that
```python
self.timezone.localize(local_time)
```
never worked because there is no attribute `localize()` in `str`. Not sure when this happened 😅 but it is fixed.

### CODE REVIEW

**Comment by dphoria** [comment]
I don't understand why it's seemingly not using the latest commit in my fork. The `diff` shown on the "Files changed" tab is not against 84a207c0e307be4e5292e062015a6768e5c89100.

**Comment by JacksonMaxfield** [comment]
Hey @dphoria I'm a bit busy, would you mind running this branch for seattle events and dumping to a JSON file then sending me that file somewhere? Or posting the JSON here on the PR.

**Comment by dphoria** [comment]
If only I had my dumb bot ready haha. The text file is in https://web.tresorit.com/l/hld2Y#muD7-qt0l42JeX0Mfqi7kw . It's for the past 4 weeks in Seattle. An example of a `Matter`:

```json
"matter": {
    "name": "Appt 01936",
    "matter_type": "Appointment (Appt)",
    "title": "Reappointment of Frank F. Alvarado III as member, Community Roots Housing Public Development Authority Governing Council, for a term to March 31, 2024.",
    "result_status": "Adopted",
    "sponsors": [
        {
            "name": "Teresa Mosqueda",
            "is_active": true,
            "router_string": null,
            "email": "Teresa.Mosqueda@seattle.gov",
            "phone": "206-684-8806",
            "website": "http://www.seattle.gov/council/mosqueda",
            "picture_uri": null,
            "seat": null,
            "roles": [
                {
                    "title": "Councilmember",
                    "body": {
                        "name": "City Council",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "138"
                    },
                    "start_datetime": 1512028800.0,
                    "end_datetime": 1514707200.0,
                    "external_source_id": "932"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Council Briefing",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "211"
                    },
                    "start_datetime": 1511856000.0,
                    "end_datetime": 1514707200.0,
                    "external_source_id": "933"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Budget Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "198"
                    },
                    "start_datetime": 1511942400.0,
                    "end_datetime": 1514707200.0,
                    "external_source_id": "934"
                },
                {
                    "title": "Chair",
                    "body": {
                        "name": "Housing, Health, Energy, and Workers\u2019 Rights Committee",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "233"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "936"
                },
                {
                    "title": "Vice Chair",
                    "body": {
                        "name": "Gender Equity, Safe Communities, New Americans, and Education Committee",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "236"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "954"
                },
                {
                    "title": "Alternate",
                    "body": {
                        "name": "Human Services, Equitable Development, and Renter Rights Committee",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "237"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "961"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Governance, Equity, and Technology Committee",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "238"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "964"
                },
                {
                    "title": "Councilmember",
                    "body": {
                        "name": "City Council",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "138"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "974"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Council Briefing",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "211"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "976"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on Citywide Mandatory Housing Affordability (MHA)",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "239"
                    },
                    "start_datetime": 1516608000.0,
                    "end_datetime": 1546243200.0,
                    "external_source_id": "999"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on Families, Education, Pre-School, and Promise Levy",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "240"
                    },
                    "start_datetime": 1519632000.0,
                    "end_datetime": 1546243200.0,
                    "external_source_id": "1008"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Budget Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "198"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "1017"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on Civic Arenas",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "232"
                    },
                    "start_datetime": 1524812400.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "1027"
                },
                {
                    "title": "Chair",
                    "body": {
                        "name": "Select Committee on the 2018 Seattle City Light Strategic Plan",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "241"
                    },
                    "start_datetime": 1526540400.0,
                    "end_datetime": 1546243200.0,
                    "external_source_id": "1030"
                },
                {
                    "title": "Vice Chair",
                    "body": {
                        "name": "Select Committee on Homelessness and Housing Affordability",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "242"
                    },
                    "start_datetime": 1534316400.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "1040"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on Citywide Mandatory Housing Affordability (MHA)",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "239"
                    },
                    "start_datetime": 1546329600.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "1054"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on the Library Levy",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "243"
                    },
                    "start_datetime": 1550044800.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "1063"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on Labor",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "190"
                    },
                    "start_datetime": 1514793600.0,
                    "end_datetime": 1577779200.0,
                    "external_source_id": "1079"
                },
                {
                    "title": "Chair",
                    "body": {
                        "name": "Finance and Housing Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "247"
                    },
                    "start_datetime": 1578297600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1117"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Governance and Education Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "248"
                    },
                    "start_datetime": 1578297600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1125"
                },
                {
                    "title": "Vice Chair",
                    "body": {
                        "name": "Land Use and Neighborhoods Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "249"
                    },
                    "start_datetime": 1578297600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1130"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Public Assets and Native Communities Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "250"
                    },
                    "start_datetime": 1578297600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1138"
                },
                {
                    "title": "Alternate",
                    "body": {
                        "name": "Sustainability and Renters' Rights Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "252"
                    },
                    "start_datetime": 1578297600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1152"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on Homelessness Strategies and Investments",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "254"
                    },
                    "start_datetime": 1578384000.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1165"
                },
                {
                    "title": "Vice Chair",
                    "body": {
                        "name": "Select Committee on Labor",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "190"
                    },
                    "start_datetime": 1578384000.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1170"
                },
                {
                    "title": "Chair",
                    "body": {
                        "name": "Select Budget Committee",
                        "is_active": true,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "198"
                    },
                    "start_datetime": 1577865600.0,
                    "end_datetime": 1640937600.0,
                    "external_source_id": "1176"
                },
                {
                    "title": "Member",
                    "body": {
                        "name": "Select Committee on Seattle Transportation Benefit District Funding",
                        "is_active": false,
                        "start_datetime": null,
                        "description": null,
                        "end_datetime": null,
                        "external_source_id": "255"
                    },
                    "start_datetime": 1593673200.0,
                    "end_datetime": 1609401600.0,
                    "external_source_id": "1191"
                }
            ],
            "external_source_id": "594"
        }
    ],
    "external_source_id": "11743"
}
```

**Comment by JacksonMaxfield** [comment]
@tohuynh and @BrianL3  would love your opinions on these `title` values.

**Review by JacksonMaxfield — APPROVED** [review]
Looks good to me!

Let's wait for To and Brian's opinions on the role titles. We may want constants??? probably not but just open for discussion.

This is just roles, did you find anything in the Legistar API for seats?

**Comment by JacksonMaxfield** [comment]
Based off of our database diagram I legitimately don't know if this can be merged without Seat additions.

https://councildataproject.org/cdp-backend/database_schema.html

**Comment by dphoria** [comment]
I tried to find information appropriate for CDP `Seat` but failed. I'll look again.

**Comment by JacksonMaxfield** [comment]
Hmmmm maybe we need to add a system to the scraper that is just a local data look up. Literally at the top of the scraper class is a bunch of constants that can add that information.

Seat info is pretty easy to maintain. We would have to do it once every two years ya know?

**Comment by JacksonMaxfield** [comment]
Wow yea, Legistar doesn't have any Seat stuff.

**Comment by dphoria** [comment]
Just for reference for Brian and To, here is an example of a `Role` in Legistar. (OfficeRecord in Legistar)

```
{
    'OfficeRecordId': 1081,
    'OfficeRecordGuid': '4C2FC0FD-F5CA-494B-A409-C3BB25E2BC44',
    'OfficeRecordLastModifiedUtc': '2019-11-27T00:06:33.073',
    'OfficeRecordRowVersion': 'AAAAAACY/D4=',
    'OfficeRecordFirstName': 'Alex',
    'OfficeRecordLastName': 'Pedersen',
    'OfficeRecordEmail': 'Alex.Pedersen@seattle.gov',
    'OfficeRecordFullName': 'Alex Pedersen',
    'OfficeRecordStartDate': '2019-11-26T00:00:00',
    'OfficeRecordEndDate': '2019-12-31T00:00:00',
    'OfficeRecordSort': 1004,
    'OfficeRecordPersonId': 661,
    'OfficeRecordBodyId': 222,
    'OfficeRecordBodyName': 'Civil Rights, Utilities, Economic Development, and Arts Committee',
    'OfficeRecordTitle': 'Alternate',
    'OfficeRecordVoteDivider': 1.0,
    'OfficeRecordExtendFlag': 0,
    'OfficeRecordMemberTypeId': 3,
    'OfficeRecordMemberType': 'Member',
    'OfficeRecordSupportNameId': None,
    'OfficeRecordSupportFullName': None,
    'OfficeRecordBodyInfo':
    {
        'BodyId': 222,
        'BodyGuid': '2196F9D6-95A7-4DA2-8D18-864FCFA190ED',
        'BodyLastModifiedUtc': '2015-12-19T01:13:46.363',
        'BodyRowVersion': 'AAAAAACynOA=',
        'BodyName': 'Civil Rights, Utilities, Economic Development, and Arts Committee',
        'BodyTypeId': 43,
        'BodyTypeName': 'Council Committee',
        'BodyMeetFlag': 1,
        'BodyActiveFlag': 0,
        'BodySort': 1000,
        'BodyDescription': 'http://www.seattle.gov/council/committees/utilities\r\n\r\n\r\nCivil Rights, Utilities, Economic Development, and Arts Committee: To provide policy direction and oversight and to deliberate and make recommendations on legislative matters relating to: civil rights issues; Seattle Public Utilities; economic development policies and programs; including small business development and support, Business Improvement Areas, workforce development, and improving access and opportunities to education and training for low- and middle-income workers, youth and communities of color; arts and cultural activities; nightlife issues; and film and music activities.',
        'BodyContactNameId': None,
        'BodyContactFullName': None,
        'BodyContactPhone': None,
        'BodyContactEmail': None,
        'BodyUsedControlFlag': 0,
        'BodyNumberOfMembers': 4,
        'BodyUsedActingFlag': 0,
        'BodyUsedTargetFlag': 0,
        'BodyUsedSponsorFlag': 0
    }
}
```

Hm... OfficeRecordMemberTypeId and OfficeRecordMemberType look... interesting???

**Review by JacksonMaxfield — CHANGES_REQUESTED** [review]
Marking as needing changes because of Seat / Role

**Comment by tohuynh** [comment]
I'm fine with using `OfficeRecordTitle` for the `role.title`, since it has a bit more specificity than `OfficeRecordMemberType`.

The Legistar website does have some info on seats (council position no .2, etc). Don't know how to get them from the API though.
image

**Comment by dphoria** [comment]
> 
> 
> The Legistar website does have some info on seats (council position no .2, etc). Don't know how to get them from the API though.

Oh thank you for that. I'll reinvestigate Legistar API data first then perhaps just scrape it. But I'm guessing it is somewhere in the Legistar data, if it's on the web site.

**Comment by JacksonMaxfield** [comment]
@dphoria I just updated and published a new `cdp-backend` version and updated the version pin in this repo, be sure to merge `main` into your branch. It is the `roles` are now dependents of `seat` ingestion model change.

**Comment by JacksonMaxfield** [comment]
/test-scraper get_seattle_events --from_dt=2021-08-01T00:00:00.000 --to_dt=2021-08-03T00:00:00

**Comment by dphoria** [comment]
> 
> 
> /test-scraper get_seattle_events --from_dt=2021-08-01T00:00:00.000 --to_dt=2021-08-03T00:00:00

Ugh failed

**Comment by JacksonMaxfield** [comment]
/test-scraper get_seattle_events --from_dt=2021-08-01T00:00:00.000 --to_dt=2021-08-03T00:00:00

**Comment by JacksonMaxfield** [comment]
@dphoria the bot worked like a charm. Thanks for that work!! :slightly_smiling_face: 

One request on the bot:
Can you split up the "params" into one file and the results in a JSON file. If the file is markdown I cant collapse the various parts of the JSON tree.

Onto stuff about this PR:

The results from the test returned this for a `Person`
```json
  "person": {
    "name": "M. Lorena Gonz\u00e1lez",
    "is_active": true,
    "router_string": null,
    "email": "lorena.gonzalez@seattle.gov",
    "phone": "206-684-8802",
    "website": "http://www.seattle.gov/council/gonzalez",
    "picture_uri": null,
    "seat": null,
    "roles": [
      {
        "title": "Councilmember",
        "body": {
          "name": "City Council",
          "is_active": true,
          "start_datetime": null,
          "description": null,
          "end_datetime": null,
          "external_source_id": "138"
        },
        "start_datetime": 1448438400.0,
        "end_datetime": 1514707200.0,
        "external_source_id": "812"
      },
      {
        "title": "Member",
        "body": {
          "name": "Committee on Central Waterfront, Seawall, and Alaskan Way Viaduct Replacement Program",
          "is_active": false,
          "start_datetime": null,
          "description": null,
          "end_datetime": null,
          "external_source_id": "188"
        },
        "start_datetime": 1448438400.0,
        "end_datetime": 1514707200.0,
        "external_source_id": "815"
      }
    ]
  }
```

I see the bug in the code but it isn't a bug because I forgot to remove `roles` from `Person` upstream.

File: cdp_scrapers/legistar_utils.py [hunk 1]
```diff
@@ -677,11 +946,12 @@ def get_person(self, legistar_person: Dict) -> Optional[Person]:
         return self.get_none_if_empty(
             Person(
                 email=str_simplified(legistar_person[LEGISTAR_PERSON_EMAIL]),
-                external_source_id=legistar_person[LEGISTAR_PERSON_EXT_ID],
+                external_source_id=str(legistar_person[LEGISTAR_PERSON_EXT_ID]),
                 name=str_simplified(legistar_person[LEGISTAR_PERSON_NAME]),
                 phone=phone,
                 website=str_simplified(legistar_person[LEGISTAR_PERSON_WEBSITE]),
                 is_active=bool(legistar_person[LEGISTAR_PERSON_ACTIVE]),
+                roles=self.get_roles(legistar_person[LEGISTAR_PERSON_ROLES]),
```
Comment (JacksonMaxfield) [inline cdp_scrapers/legistar_utils.py:954]: This should be removed. I will fix upstream but `roles` should be under `seat`

In the future this should be `seat=Seat(..., roles=self.get_roles(...))`

and by "future" i mean in like 20 minutes when I release v3.0.0.dev15

**Comment by JacksonMaxfield** [comment]
Alright @dphoria should be good to update to `cdp-backend~=3.0.0.dev15` for latest ingestion models. `roles` are now entirely nested under `seat`

**Comment by JacksonMaxfield** [comment]
/test-scraper get_seattle_events --from_dt=2021-08-01T00:00:00.000 --to_dt=2021-08-03T00:00:00

**Comment by JacksonMaxfield** [comment]
From CI log, just need to move the `roles` to under the `Seat` object not the `Person` object.

**Comment by dphoria** [comment]
`Seat.roles` will be added along with `Person.seat` after #31 and #32 .

**Comment by dphoria** [comment]
/test-scraper get_seattle_events --from_dt=2021-08-01T00:00:00.000 --to_dt=2021-08-03T00:00:00

**Review by JacksonMaxfield — APPROVED** [review]
Looks good to me! Thanks!

**Pull request merged**